x/upgrade/abci.go

```diff
@@ -9,30 +9,39 @@ import (
 )
 
 // BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If skip upgrade flag is set, it will skip and clear the upgrade plan
 // If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
 // If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
 //
-// The purpose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
-func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock, skipUpgradeHeightArray []int64) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
 
 	if plan.ShouldExecute(ctx) {
+		//To make sure clear upgrade is executed at the same block
+		if k.Contains(skipUpgradeHeightArray, ctx.BlockHeight()) {
+			// If skip upgrade has been set, we clear the upgrade plan
+			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %d: %s", plan.Name, plan.Height, plan.Info)
+			ctx.Logger().Info(skipUpgradeMsg)
+			k.ClearUpgradePlan(ctx)
```

Comment:
This only needs to be called once, no?

Reply:
I am not sure if you are asking about ClearUpgradePlan here, please confirm. ClearUpgradePlan only clears the plan at that particular height

Reply:
Yes, `ClearUpgradePlan`. But it does look like it's only being called once. Thanks!

Reply:
> Looking good @anilCSE -- left a few more comments 
> 
> Question: `skipUpgradeHeights` are provided to the module from user CLI input, no? This array/map dictates the control flow of `BeginBlock` (in that it clears the plan). What if different nodes set (or dont set at all) different values of `skipUpgradeHeights`? Is this OK?
Case - 1:
We're checking this only if there's an upgrade plan to be executed at that height, so it doesn't care if a set of heights which don't have upgrade scheduled are passed. So if user passes bunch of random heights,there will be no effect. 
Case - 2:
This implemented solution can be used only to skip upgrades which have passed a proposal. So it's in validators intrest that they chose to upgrade at that height or not based on how social consensus on a particular update is. If they chose otherwise, they'll not be able to sync or sign the blocks after that height. Either the ways, it shouldn't be a problem

Reply:
Added a godoc to make it clear